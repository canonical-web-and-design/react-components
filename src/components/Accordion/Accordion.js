import classNames from "classnames";
import PropTypes from "prop-types";
import React, { useState } from "react";

import AccordionSection from "../AccordionSection";

const generateSections = (sections, setExpanded, expanded) =>
  sections.map(({ content, title }, i) => (
    <AccordionSection
      content={content}
      expanded={expanded}
      key={i}
      setExpanded={setExpanded}
      title={title}
    />
  ));

const Accordion = ({ className, sections, ...props }) => {
  const [expanded, setExpanded] = useState();
  return (
    <aside
      className={classNames(className, "p-accordion")}
      {...props}
      role="tablist"
      aria-multiselectable="true"
    >
      <ul className="p-accordion__list">
        {generateSections(sections, setExpanded, expanded)}
      </ul>
    </aside>
  );
};

Accordion.propTypes = {
  /**
   * Optional classes applied to the parent element.
   */
  className: PropTypes.string,
  /**
   * An array of sections and content.
   */
  sections: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The title of the section.
       */
      title: PropTypes.string,
      /**
       * The content of the section.
       */
      content: PropTypes.node
    }).isRequired
  )
};

export default Accordion;
