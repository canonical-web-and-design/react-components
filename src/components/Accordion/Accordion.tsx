import classNames from "classnames";
import PropTypes from "prop-types";
import React, { useState, HTMLProps } from "react";

import AccordionSection from "../AccordionSection";
import type {
  AccordionHeadings,
  AccordionSectionProps,
} from "../AccordionSection";

const generateSections = (
  sections: AccordionSectionProps & { key?: string | number }[],
  setExpanded,
  expanded,
  titleElement
) =>
  sections.map(({ key, ...props }, i) => (
    <AccordionSection
      expanded={expanded}
      key={key || i}
      sectionKey={key?.toString()}
      setExpanded={setExpanded}
      titleElement={titleElement}
      {...props}
    />
  ));

type Props = {
  sections: typeof AccordionSection[];
  className?: string;
  expanded?: string;
  externallyControlled?: boolean;
  onExpandedChange?: (id, title: string) => void;
  titleElement?: AccordionHeadings;
} & HTMLProps<HTMLElement>;

const Accordion = ({
  className,
  expanded,
  externallyControlled,
  onExpandedChange,
  sections,
  titleElement,
  ...asideProps
}: Props): JSX.Element => {
  const [expandedSection, setExpandedSection] = useState(expanded);
  const setExpanded = (id, title) => {
    setExpandedSection(id);
    onExpandedChange && onExpandedChange(id, title);
  };
  return (
    <aside
      className={classNames(className, "p-accordion")}
      {...asideProps}
      role="tablist"
      aria-multiselectable="true"
    >
      <ul className="p-accordion__list">
        {generateSections(
          sections,
          setExpanded,
          externallyControlled ? expanded : expandedSection,
          titleElement
        )}
      </ul>
    </aside>
  );
};

Accordion.propTypes = {
  /**
   * Optional classes applied to the parent element.
   */
  className: PropTypes.string,
  /**
   * An optional value to set the expanded section. The value must match a
   * section key. This value will only set the expanded section on first render
   * if externallyControlled is not set to `true`.
   */
  expanded: PropTypes.string,
  /**
   * Whether the expanded section will be controlled via external state.
   */
  externallyControlled: PropTypes.bool,
  /**
   * An array of sections and content.
   */
  sections: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The content of the section.
       */
      content: PropTypes.node,
      /**
       * An optional key for the section component. It will also be
       * used to track which section is selected.
       */
      key: PropTypes.string,
      /**
       * An optional click event when the title is clicked.
       */
      onTitleClick: PropTypes.func,
      /**
       * The title of the section.
       */
      title: PropTypes.string,
    }).isRequired
  ),
  /**
   * Optional function that is called when the expanded section is changed.
   * The function is provided the section title or null.
   */
  onExpandedChange: PropTypes.func,
  /**
   * Optional string describing heading element that should be used for the section titles.
   */
  titleElement: PropTypes.oneOf(["h2", "h3", "h4", "h5", "h6"]),
};

export default Accordion;
