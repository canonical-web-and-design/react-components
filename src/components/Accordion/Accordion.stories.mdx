import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs, object } from "@storybook/addon-knobs";
import { useState } from "react";
import Accordion from "./Accordion";

<Meta title="Accordion" component={Accordion} decorators={[withKnobs]} />

### Accordion

This is a [React](https://reactjs.org/) component for the Vanilla [Accordion](https://docs.vanillaframework.io/patterns/accordion/).

The sidebar accordion, used in listing pages or as navigation, can hold multiple navigation items or to be used as a filter for content.

### Props

<Props of={Accordion} />

### Default

<Preview>
  <Story name="Default">
    <Accordion
      sections={[
        {
          title: "Advanced topics",
          content: (
            <>
              <p>Charm bundles</p>
              <p>Machine authentication</p>
              <p>Migrating models</p>
              <p>Using storage</p>
              <p>Working with actions</p>
              <p>Working with resources</p>
              <p>Cloud image metadata</p>
              <p>Tools</p>
            </>
          )
        },
        {
          title: "Networking",
          content: (
            <>
              <p>Working offline</p>
              <p>Fan container networking</p>
              <p>Network spaces</p>
            </>
          )
        },
        {
          title: "Miscellaneous",
          content: (
            <>
              <p>Juju GUI</p>
              <p>CentOS support</p>
              <p>Collecting Juju metrics</p>
            </>
          )
        }
      ]}
    />
  </Story>
</Preview>

### External state

The expanded accordion section can be controlled by external state.

<Preview>
  <Story name="External state">
    {() => {
      const [expandedSection, setExpandedSection] = useState();
      return (
        <Accordion
          expanded={expandedSection}
          externallyControlled
          onExpandedChange={setExpandedSection}
          sections={[
            {
              title: "Advanced topics",
              key: "advanced-topics",
              content: (
                <>
                  <p>Charm bundles</p>
                  <p>Machine authentication</p>
                  <p>Migrating models</p>
                  <p>Using storage</p>
                  <p>Working with actions</p>
                  <p>Working with resources</p>
                  <p>Cloud image metadata</p>
                  <p>Tools</p>
                </>
              )
            },
            {
              title: "Networking",
              key: "networking",
              content: (
                <>
                  <p>Working offline</p>
                  <p>Fan container networking</p>
                  <p>Network spaces</p>
                </>
              )
            },
            {
              title: "Miscellaneous",
              key: "miscellaneous",
              content: (
                <>
                  <p>Juju GUI</p>
                  <p>CentOS support</p>
                  <p>Collecting Juju metrics</p>
                </>
              )
            }
          ]}
        />
      );
    }}
  </Story>
</Preview>


### Headings

`titleElement` prop can be used to define heading element for section titles.

<Preview>
  <Story name="Headings">
    <Accordion
      sections={[
        {
          title: "Advanced topics",
          content: (
            <>
              <p>Charm bundles</p>
              <p>Machine authentication</p>
              <p>Migrating models</p>
              <p>Using storage</p>
              <p>Working with actions</p>
              <p>Working with resources</p>
              <p>Cloud image metadata</p>
              <p>Tools</p>
            </>
          )
        },
        {
          title: "Networking",
          content: (
            <>
              <p>Working offline</p>
              <p>Fan container networking</p>
              <p>Network spaces</p>
            </>
          )
        },
        {
          title: "Miscellaneous",
          content: (
            <>
              <p>Juju GUI</p>
              <p>CentOS support</p>
              <p>Collecting Juju metrics</p>
            </>
          )
        }
      ]}
      titleElement="h3"
    />
  </Story>
</Preview>
