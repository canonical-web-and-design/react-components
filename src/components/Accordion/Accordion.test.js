import { mount, shallow } from "enzyme";
import React from "react";

import Accordion from "./Accordion";

jest.mock("uuid/v4", () =>
  jest.fn(() => "00000000-0000-0000-0000-000000000000")
);

describe("Accordion ", () => {
  it("renders", () => {
    const wrapper = shallow(
      <Accordion
        sections={[
          {
            title: "Advanced topics",
            content: "test content"
          },
          {
            title: "Networking",
            content: <>More test content</>
          }
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("can call a function when a section is expanded", () => {
    const onExpandedChange = jest.fn();
    const wrapper = mount(
      <Accordion
        onExpandedChange={onExpandedChange}
        sections={[
          {
            title: "Advanced topics",
            content: "test content"
          },
          {
            title: "Networking",
            content: <>More test content</>
          }
        ]}
      />
    );
    const title = wrapper.find(".p-accordion__tab").at(0);
    title.simulate("click");
    expect(onExpandedChange).toHaveBeenCalledWith("Advanced topics");
    // Clicking the title again should close the accordion section.
    title.simulate("click");
    expect(onExpandedChange).toHaveBeenCalledWith(null);
  });

  it("can add additional classes", () => {
    const wrapper = shallow(
      <Accordion className="extra-class" sections={[]} />
    );
    const className = wrapper.prop("className");
    expect(className.includes("p-accordion")).toBe(true);
    expect(className.includes("extra-class")).toBe(true);
  });
});
