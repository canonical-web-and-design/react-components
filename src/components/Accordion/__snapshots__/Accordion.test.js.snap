// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion  renders 1`] = `
<aside
  aria-multiselectable="true"
  className="p-accordion"
  role="tablist"
>
  <ul
    className="p-accordion__list"
  >
    <AccordionSection
      content="test content"
      key="0"
      setExpanded={[Function]}
      title="Advanced topics"
    />
    <AccordionSection
      content={
        <React.Fragment>
          More test content
        </React.Fragment>
      }
      key="1"
      setExpanded={[Function]}
      title="Networking"
    />
  </ul>
</aside>
`;

exports[`Accordion  renders headings for titles 1`] = `
<aside
  aria-multiselectable="true"
  className="p-accordion"
  role="tablist"
>
  <ul
    className="p-accordion__list"
  >
    <AccordionSection
      content="test content"
      key="0"
      setExpanded={[Function]}
      title="Advanced topics"
      titleElement="h4"
    />
    <AccordionSection
      content={
        <React.Fragment>
          More test content
        </React.Fragment>
      }
      key="1"
      setExpanded={[Function]}
      title="Networking"
      titleElement="h4"
    />
  </ul>
</aside>
`;
